leftwm-core/src/handlers/focus_handler.rs

```diff
@@ -96,12 +93,97 @@ impl State {
 
         // Unfocus last window if the target tag is empty
         if let Some(window) = self.focus_manager.window(&self.windows) {
-            if !window.tags.contains(tag) {
+            if window.tag != Some(*tag) {
                 self.unfocus_current_window();
             }
         }
     }
 
+    /// Focuses the workspace containing a given point.
+    pub fn focus_workspace_with_point(&mut self, x: i32, y: i32) {
+        let focused_id = match self.focus_manager.workspace(&self.workspaces) {
+            Some(fws) => fws.id,
+            None => None,
+        };
+        if let Some(ws) = self
+            .workspaces
+            .iter()
+            .find(|ws| ws.contains_point(x, y) && ws.id != focused_id)
+            .cloned()
+        {
+            self.focus_workspace(&ws);
+        }
+    }
+
+    /// Focuses the window containing a given point.
+    pub fn focus_window_with_point(&mut self, x: i32, y: i32) {
+        let handle_found: Option<WindowHandle> = self
+            .windows
+            .iter()
+            .filter(|x| x.can_focus())
+            .find(|w| w.contains_point(x, y))
+            .map(|w| w.handle);
+        match handle_found {
+            Some(found) => self.focus_window(&found),
+            //backup plan, move focus closest window in workspace
+            None => self.focus_closest_window(x, y),
+        }
+    }
+
+    /// Validates that the given window is focused.
+    pub fn validate_focus_at(&mut self, handle: &WindowHandle) {
+        // If the window is already focused do nothing.
+        if let Some(current) = self.focus_manager.window(&self.windows) {
+            if &current.handle == handle {
+                return;
+            }
+        }
+        // Focus the window only if it is also focusable.
+        if self
+            .windows
+            .iter()
+            .any(|w| w.can_focus() && &w.handle == handle)
+        {
+            self.focus_window(handle);
+        }
+    }
+
+    // Helper function.
+
+    fn focus_closest_window(&mut self, x: i32, y: i32) {
+        let ws = match self.workspaces.iter().find(|ws| ws.contains_point(x, y)) {
+            Some(ws) => ws,
+            None => return,
+        };
+        let mut dists: Vec<(i32, &Window)> = self
+            .windows
+            .iter()
+            .filter(|x| ws.is_managed(x) && x.can_focus())
+            .map(|w| (distance(w, x, y), w))
+            .collect();
+        dists.sort_by(|a, b| (a.0).cmp(&b.0));
+        if let Some(first) = dists.get(0) {
```

Comment:
and this, but here it might actually make sense to replace anyways, right?
```suggestion
        if let Some(first) = dists.first() {
```